### PR TITLE
Properly handle shorthands in destructure patterns when renaming

### LINKED ISSUE #6548: Renaming a struct initializer to match the field name should use shorthand record syntax
I have code like this:

```rust
let inner = unimplemented!();
Item { kind: inner, ..unimplemented!() }
```

When I rename `inner` to `kind`, I expect this to turn into `Item { kind, ..unimplemented!() }`.
Instead, it turns into `Item { kind: kind, ..unimplemented!() }`, and I have to hit ctrl+. again to get the short-hand syntax.

### PR DESCRIPTION
**Pull request by Veykril**
Fixes #6548 and #6551.

### CODE REVIEW

File: crates/ide/src/references/rename.rs [hunk 1]
```diff
@@ -122,12 +125,49 @@ fn source_edit_from_reference(reference: Reference, new_name: &str) -> SourceFil
...
         }
     };
-    SourceFileEdit { file_id, edit: TextEdit::replace(range, replacement_text) }
+    SourceFileEdit {
+        file_id: reference.file_range.file_id,
+        edit: TextEdit::replace(range, replacement_text),
+    }
+}
+
+fn edit_text_range_for_record_field_expr_or_pat(
```
Comment (SomeoneToIgnore) [inline crates/ide/src/references/rename.rs:143]: I've came up with an alternative version with no `mut` and hence less indirection.
The naming is not perfect, but if you like the approach on general, use it 🙂 

<details>

```rust
fn edit_text_range_for_record_field_expr_or_pat(
    sema: &Semantics<RootDatabase>,
    file_range: FileRange,
    new_name: &str,
) -> TextRange {
    let source_file = sema.parse(file_range.file_id);
    let file_syntax = source_file.syntax();
    let original_range = file_range.range;

    syntax::algo::find_node_at_range::<ast::RecordExprField>(file_syntax, original_range)
        .and_then(|field_expr| match field_expr.expr().and_then(|e| e.name_ref()) {
            Some(name) if &name.to_string() == new_name => Some(field_expr.syntax().text_range()),
            _ => None,
        })
        .or_else(|| {
            syntax::algo::find_node_at_range::<ast::RecordPatField>(file_syntax, original_range)
                .and_then(|field_pat| match field_pat.pat() {
                    Some(ast::Pat::IdentPat(pat))
                        if pat.name().map(|n| n.to_string()).as_deref() == Some(new_name) =>
                    {
                        Some(field_pat.syntax().text_range())
                    }
                    _ => None,
                })
        })
        .unwrap_or(original_range)
}
```  

</details>

Comment (Veykril) [reply crates/ide/src/references/rename.rs:143]: This looks a lot better, thanks. Didn't like the the `mut` either but I imagined using option adaptors would look a lot worse here. Nice to know it doesn't 🙂

File: crates/ide/src/references/rename.rs [hunk 2]
```diff
@@ -1094,6 +1134,104 @@ impl Foo {
...
+fn foo(foo: Foo) {
+    let Foo { i: bar } = foo;
+    let _ = bar;
+}
+"#,
+        );
+    }
+
+    #[test]
+    fn test_struct_field_destructure_into_shorthand() {
```
[the remark this answers is not in the document]
Comment (Veykril) [reply crates/ide/src/references/rename.rs:1217]: Regarding the nit you mean as in merging most of the before_fixtures of the difference cases to reuse them behind a static?

And yes, testing for parameter destructuring also seems like a good idea

[the remark this answers is not in the document]
Comment (Veykril) [reply crates/ide/src/references/rename.rs:1217]: Hm I don't think we can put the tests together though, as they require slightly different fixtures and/or different `<|>` cursor positions.

[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply crates/ide/src/references/rename.rs]: Oh, never mind then, I din't think it's worth supporting the multiple caret tests for this case.

File: crates/ide/src/references/rename.rs [hunk 3]
```diff
@@ -126,6 +126,7 @@ fn source_edit_from_reference(
             TextRange::new(reference.file_range.range.end(), reference.file_range.range.end())
         }
         ReferenceKind::RecordFieldExprOrPat => {
+            mark::hit!(test_rename_field_expr_pat);
```
Comment (SomeoneToIgnore) [inline crates/ide/src/references/rename.rs:129]: 👍

File: crates/syntax/src/ast/expr_ext.rs [hunk 4]
```diff
@@ -22,6 +22,18 @@ impl ast::Expr {
             _ => false,
         }
     }
+
+    pub fn name_ref(&self) -> Option<ast::NameRef> {
```
Comment (matklad) [inline crates/syntax/src/ast/expr_ext.rs:26]: The same thing here -- this methong is non-primitives -- it's a convenience function which reaches deep into the node. We *should* have such functions, but not as inherent methods on AST nodes.

Comment (matklad) [reply crates/syntax/src/ast/expr_ext.rs:26]: the same -- see my recent comment on another PR :-)

**Pull request merged**